diffkemp/simpll/passes/FieldAccessFunctionGenerator.cpp

```diff
@@ -0,0 +1,191 @@
+    auto &CalledFuns = mam.getResult<CalledFunctionsAnalysis>(Mod, Main);
+
+    for (Function &Fun : Mod) {
+        if (!ModOther->getFunction(Fun.getName()))
+            continue;
+
+        if (CalledFuns.find(&Fun) == CalledFuns.end())
+            continue;
+
+        std::vector<Instruction *> InstStack;
+
+        // This stores the debug location for the currently processed stack.
+        // This enables the stopping of the building of the stack when an
+        // instruction with a different DILocation is reached.
+        DILocation *DIL;
+
+        // This is a FSM-like algorithm that gathers together groups of
+        // instructions beginning with a GEP and containing GEPs and casts.
+        // (They also have to have the same debug location, i.e. correspond to
+        // a single field access in the C source code).
+        // A vector/stack is generated from each of them and the processStack
+        // function is called to further process the instructions.
+        for (auto &BB : Fun) {
+            for (auto &Inst : BB) {
+                if (isa<GetElementPtrInst>(Inst) && InstStack.empty()) {
+                    // Possible start of field access block.
+                    // Check if GEP has constant indices (this pass could
+                    // technically support all GEPs, but this is not implemented
+                    // yet)
+                    if (dyn_cast<GetElementPtrInst>(&Inst)->
+                            hasAllConstantIndices()) {
+                        DIL = Inst.getDebugLoc().get();
+                        InstStack.push_back(&Inst);
+                    }
+                } else if ((isa<GetElementPtrInst>(Inst) ||
+                           isa<CastInst>(Inst)) && !InstStack.empty()) {
+                   if (Inst.getDebugLoc().get() != DIL) {
+                       // Stop building stack.
+                       processStack(InstStack, Mod);
+                   } else {
+                       // Add instruction to stack.
+                       InstStack.push_back(&Inst);
+                   }
+                } else {
+                    // Wrong instruction type. Stop building stack.
+                    processStack(InstStack, Mod);
+                }
+            }
+            // Avoid processing instruction groups that contain instructions in
+            // two or more basic blocks.
+            processStack(InstStack, Mod);
+        }
+    }
+
+    return PreservedAnalyses();
+}
+
+/// First checks whether the instructions in the stack can be moved away from
+/// the original function without causing an error, then creates the abstraction
+/// function, moves the instructions into it, modify the input and output of the
+/// block to match the argument and returns value of the function and inserts
+/// the call to the newly create abstraction instead of the instruction block to
+/// the original function.
+void FieldAccessFunctionGenerator::processStack(
```

Comment:
I'd suggest a slightly different implementation:
- make `Stack` constant: this function should not modify it except for clearing it at the end - this can be done in the caller function in a better place (see other comment above)
- since you don't have to clear it here, you can get rid of `ValidUserSet`, `allUsersValid`, and `allOperandsValid` variables (more details in the comments below).